### PR TITLE
bpo-32769: A new take on annotations/type hinting glossary entries

### PR DESCRIPTION
**Pull request by andresdelfino**
https://bugs.python.org/issue32769

### CODE REVIEW

**Comment by andresdelfino** [comment]
@ilevkivskyi I would really appreciate any input, when you have some spare time of course.

File: Doc/glossary.rst [hunk 1]
```diff
@@ -40,16 +40,18 @@ Glossary
       ABCs with the :mod:`abc` module.
 
    annotation
-      A metadata value associated with a global variable, a class attribute or a
-      function or method parameter or return value, that stores a
-      :term:`type hint`.
+      A metadata value associated with a variable, a class attribute or a
```
Comment (ilevkivskyi) [inline Doc/glossary.rst]: I may guess "value" is what @gvanrossum didn't like, and I agree with him.

File: Doc/glossary.rst [hunk 2]
```diff
@@ -1053,21 +1082,37 @@ Glossary
...
       When annotating variables, assignment is optional::
 
          class C:
-             field: int
+             field: 'annotation'
+
+      Annotating local variables will cause the interpreter to always make them
+      locals::
+
+         def f():
```
Comment (ilevkivskyi) [inline Doc/glossary.rst]: I think there are too much details here, it is glossary, not a spec.

File: Doc/glossary.rst [hunk 3]
```diff
@@ -1033,17 +1036,43 @@ Glossary
...
+
+         def broadcast_message(message: str, servers: List[Server]) -> None:
+             ...
+
+      is equivalent to::
+
+         def broadcast_message(
+                 message: str,
+                 servers: List[Tuple[Tuple[str, int], Dict[str, str]]]) -> None:
+             ...
```
Comment (ilevkivskyi) [inline Doc/glossary.rst]: I would also make this example shorter.

**Review by ilevkivskyi — COMMENTED** [review]
Sorry, I lost track a bit, here are few superficial comments. I would rather ask @gvanrossum to take a look at wording here (if he has bandwidth of course).

File: Doc/glossary.rst (hunk 1 above)
Comment (andresdelfino) [reply Doc/glossary.rst]: Yes, I guess this will be the most difficult part to get right 😅

File: Doc/glossary.rst [hunk 4]
```diff
@@ -1039,8 +1040,8 @@ Glossary
    type alias
       A synonym for a type, created by assigning the type to an identifier.
 
-      Type aliases are useful for simplifying complex
-      :term:`type hints <type hint>`.  For example ::
+      Type aliases are useful for simplifying :term:`type hints <type hint>`.
+      For example ::
```
Comment (gvanrossum) [inline Doc/glossary.rst]: Honestly this example would look better if it read "For example [bad code] could be made more readable like this [good code]" (currently it shows [good code] first).

File: Doc/glossary.rst [hunk 5]
```diff
@@ -1081,7 +1082,7 @@ Glossary
    variable annotation
       An :term:`annotation` of a variable, or a class attribute.
```
Comment (gvanrossum) [inline Doc/glossary.rst]: This should not have a comma.

File: Doc/glossary.rst [hunk 6]
```diff
@@ -40,16 +40,18 @@ Glossary
       ABCs with the :mod:`abc` module.
 
    annotation
-      A metadata value associated with a global variable, a class attribute or a
-      function or method parameter or return value, that stores a
-      :term:`type hint`.
+      A label with no defined semantics associated with a variable, a class
```
Comment (gvanrossum) [inline Doc/glossary.rst]: I think "no defined semantics" is too strong -- IMO it should say that by convention annotations are usually used for type hints (i.e. as type annotations).

Also I'm not too excited about the word "label" here but I agree it's better than "metadata value".

File: Doc/glossary.rst [hunk 7]
```diff
@@ -191,7 +193,7 @@ Glossary
       A variable defined in a class and intended to be modified only at
       class level (i.e., not in an instance of the class).
 
-      Class variables can be specified as such through
+      Variables can be specified as expected to be class variables through
       :term:`type hints <type hint>`.
```
Comment (gvanrossum) [inline Doc/glossary.rst]: If you're going to be this specific I recommend calling out `ClassVar` by name.

File: Doc/glossary.rst [hunk 8]
```diff
@@ -388,19 +390,21 @@ Glossary
...
 
    function annotation
-      An :term:`annotation` of a function, or a method.
+      An :term:`annotation` of a function, or a method parameters and return
+      value.
 
-      For example, this function has its parameters annotated as taking
-      :class:`int` arguments and its return value annotated as being an
-      :class:`int` as well::
+      Function annotations can be
```
Comment (gvanrossum) [inline Doc/glossary.rst]: can be -> are usually used for

File: Doc/glossary.rst [hunk 9]
```diff
@@ -388,19 +390,21 @@ Glossary
...
+      :term:`type hints <type hint>`: this function is expected to take two
```
Comment (gvanrossum) [inline Doc/glossary.rst]: Insert "for example" before "this"

File: Doc/glossary.rst [hunk 10]
```diff
@@ -1033,17 +1037,40 @@ Glossary
...
    type hint
-      A specification about the expected type for a global variable, class
-      variable, function or method parameter or return value.
+      An :term:`annotation` that specifies the expected type for a variable, class
+      attribute, function or method parameter or return value.
 
-      While type hints are optional and are not enforced by Python when used,
-      they are useful for static type analysis tools, and aid IDEs on code
+      Type hints are optional and are not enforced by Python but
+      they are useful to static type analysis tools, and aid IDEs on code
```
Comment (gvanrossum) [inline Doc/glossary.rst]: on -> with

File: Doc/glossary.rst [hunk 11]
```diff
@@ -1033,17 +1037,40 @@ Glossary
...
       completion and refactoring.
 
-      Type hints are stored in :term:`annotations <annotation>`.
+      Type hints of global variables, class attributes, functions and methods
+      parameters and return values, but not local variables, can be accessed using
+      :func:`typing.get_type_hints`.
 
-      See also :pep:`483` which describe this functionality.
+      See :mod:`typing` and :pep:`483`, which describe this functionality.
```
Comment (gvanrossum) [inline Doc/glossary.rst]: I think this should be PEP 484. (PEP 483 also describes type hints but is much less precise and rarely referenced.)

File: Doc/glossary.rst [hunk 12]
```diff
@@ -40,16 +40,18 @@ Glossary
       ABCs with the :mod:`abc` module.
 
    annotation
-      A metadata value associated with a global variable, a class attribute or a
-      function or method parameter or return value, that stores a
-      :term:`type hint`.
+      A label with no defined semantics associated with a variable, a class
+      attribute or a function or method parameter or return value.
```
Comment (gvanrossum) [inline Doc/glossary.rst]: Please drop "or method" here and everywhere else that you wrote it -- a method is defined as a kind of function already, and the phrase "function or method parameter or return value" is too awkward.

File: Doc/glossary.rst [hunk 13]
```diff
@@ -40,16 +40,18 @@ Glossary
...
 
-      Annotations are stored in the :attr:`__annotations__` special attribute
-      of a module (when annotating a global variable), class (when annotating
-      one of its attributes) or function or method (when annotating a parameter or a
-      return value) and can be accessed using :func:`typing.get_type_hints`.
+      Annotations of local variables cannot be accesed at runtime, but
+      annotations of global variables, class attributes, functions and methods
+      parameters and return values are stored in the :attr:`__annotations__`
```
Comment (gvanrossum) [inline Doc/glossary.rst]: I'd write "and functions" instead of "functions and methods parameters and return values". Or if you really feel that it is worth repeating the "parameters and return values" then at least write "function parameters and return values", i.e. no "s".

File: Doc/glossary.rst [hunk 14]
```diff
@@ -40,16 +40,18 @@ Glossary
...
+      special attribute of modules, classes, functions and methods,
+      respectively.
 
-      See :pep:`484` and :pep:`526` which describe this functionality.
+      Annotations can be :term:`type hints <type hint>`.
```
Comment (gvanrossum) [inline Doc/glossary.rst]: I'd replace "can be" with "are usually used for"

File: Doc/glossary.rst [hunk 15]
```diff
@@ -388,19 +390,21 @@ Glossary
       and the :ref:`function` section.
 
    function annotation
-      An :term:`annotation` of a function, or a method.
+      An :term:`annotation` of a function, or a method parameters and return
```
Comment (gvanrossum) [inline Doc/glossary.rst]: I'd say "An annotation of a function parameter or return type." (plus markup on "annotation")

File: Doc/glossary.rst [hunk 16]
```diff
@@ -1033,17 +1037,40 @@ Glossary
       :attr:`~instance.__class__` attribute or can be retrieved with
       ``type(obj)``.
 
+   type alias
+      A synonym for a type, created by assigning the type to an identifier.
+
+      Type aliases are useful for simplifying :term:`type hints <type hint>`.
+      For example ::
```
Comment (gvanrossum) [inline Doc/glossary.rst]: I'd write the bad example first and the improved example next, with words such as "For example:: [bad example] can be rewritten with more clarity as:: [good example]"

File: Doc/glossary.rst [hunk 17]
```diff
@@ -1033,17 +1037,40 @@ Glossary
...
+                 colors: List[Tuple[int, int, int]]) -> List[Tuple[int, int, int]]:
+             pass
+
+      See :mod:`typing` and :pep:`484`, which describe this functionality.
+
    type hint
-      A specification about the expected type for a global variable, class
-      variable, function or method parameter or return value.
+      An :term:`annotation` that specifies the expected type for a variable, class
+      attribute, function or method parameter or return value.
```
Comment (gvanrossum) [inline Doc/glossary.rst]: Rewrite as "[...] for a variable, a class attribute, or a function parameter or return value." (The repetition of "a" makes it clear that the last "or" binds more tightly than the other.)

File: Doc/glossary.rst [hunk 18]
```diff
@@ -1033,17 +1037,40 @@ Glossary
...
 
-      While type hints are optional and are not enforced by Python when used,
-      they are useful for static type analysis tools, and aid IDEs on code
+      Type hints are optional and are not enforced by Python but
+      they are useful to static type analysis tools, and aid IDEs on code
       completion and refactoring.
 
-      Type hints are stored in :term:`annotations <annotation>`.
+      Type hints of global variables, class attributes, functions and methods
```
Comment (gvanrossum) [inline Doc/glossary.rst]: functions and methods -> function

File: Doc/glossary.rst [hunk 19]
```diff
@@ -1053,21 +1080,23 @@ Glossary
...
 
-      For example, this variable is annotated as taking :class:`int` values::
+      When annotating variables or class attributes, assignment is optional ::
 
-         count: int = 0
+         class C:
+             field: 'annotation'
 
-      When annotating variables, assignment is optional::
+      Variable annotations can be
```
Comment (gvanrossum) [inline Doc/glossary.rst]: can be -> are usually used for

File: Doc/glossary.rst [hunk 20]
```diff
@@ -1053,21 +1080,23 @@ Glossary
       :func:`bytes.splitlines` for an additional use.
 
    variable annotation
-      An :term:`annotation` of a global variable, or a class attribute.
+      An :term:`annotation` of a variable, or a class attribute.
 
-      For example, this variable is annotated as taking :class:`int` values::
+      When annotating variables or class attributes, assignment is optional ::
```
Comment (gvanrossum) [inline Doc/glossary.rst]: I'd say "a variable or a class attribute"

File: Doc/glossary.rst (hunk 7 above)
Comment (andresdelfino) [reply Doc/glossary.rst]: I agree that this was getting too specific. I left the class variable definition, though, because I think it's useful. Let me know if you think it is not and I'll remove it.

File: Doc/glossary.rst (hunk 6 above)
Comment (andresdelfino) [reply Doc/glossary.rst]: What about "syntactic construct"? Would that be too wordy?

File: Doc/glossary.rst [hunk 21]
```diff
@@ -40,18 +40,17 @@ Glossary
...
-      special attribute of modules, classes, functions and methods,
+      annotations of global variables, class attributes, and functions
+      are stored in the :attr:`__annotations__`
+      special attribute of modules, classes, and functions,
       respectively.
 
-      Annotations can be :term:`type hints <type hint>`.
-
-      See :pep:`484` and :pep:`526`, which describe this functionality.
+      See :term:`variable annotation`, :term:`function annotation`, :pep:`484` and :pep:`526`, which describe this functionality.
```
Comment (gvanrossum) [inline Doc/glossary.rst]: Please break this long line.

File: Doc/glossary.rst [hunk 22]
```diff
@@ -390,12 +384,12 @@ Glossary
...
-      An :term:`annotation` of a function, or a method.
+      An :term:`annotation` of a function parameter or return value.
 
-      Function annotations can be
-      :term:`type hints <type hint>`: this function is expected to take two
+      Function annotations are usually used for
+      :term:`type hints <type hint>`: for example this function is expected to take two
       :class:`int` arguments and is also expected to have an :class:`int`
-      return value::
+      return value ::
```
Comment (gvanrossum) [inline Doc/glossary.rst]: Why is there a space before the `::` ?

File: Doc/glossary.rst [hunk 23]
```diff
@@ -1039,40 +1033,39 @@ Glossary
...
 
       Type hints are optional and are not enforced by Python but
-      they are useful to static type analysis tools, and aid IDEs on code
+      they are useful to static type analysis tools, and aid IDEs with code
       completion and refactoring.
 
-      Type hints of global variables, class attributes, function or method
-      parameter or return value, but not local variables, can be accessed using
+      Type hints of global variables, class attributes, and functions
+      , but not local variables, can be accessed using
```
Comment (gvanrossum) [inline Doc/glossary.rst]: Move comma to preceding line.

File: Doc/glossary.rst [hunk 24]
```diff
@@ -40,18 +40,17 @@ Glossary
       ABCs with the :mod:`abc` module.
 
    annotation
-      A metadata value associated with a variable, a class attribute or a
-      function or method parameter or return value, with no defined semantics.
+      A label associated with a variable, a class
+      attribute or a function parameter or return value,
+      used, by convention, for :term:`type hints <type hint>`.
```
Comment (gvanrossum) [inline Doc/glossary.rst]: there's no need for commas around "by convention".

**Review by gvanrossum — COMMENTED** [review]
ALmost there.

File: Doc/glossary.rst (hunk 22 above)
Comment (andresdelfino) [reply Doc/glossary.rst]: I thought it was custom 😅

**Review by gvanrossum — APPROVED** [review]

**Review by ilevkivskyi — APPROVED** [review]

**Comment by ilevkivskyi** [comment]
@andresdelfino Could you please make manual backports of these changes? (Make PRs against 3.6 and 3.7 branches).

**Comment by andresdelfino** [comment]
Yes, I was working on them already :)
